core/deadline.go

```diff
@@ -45,7 +47,119 @@ type Deadliner interface {
 	C() <-chan Duty
 }
 
+// Deadline implements the Deadliner interface.
+type Deadline struct {
+	mu           sync.Mutex
+	genesisTime  time.Time
+	slotDuration time.Duration
+
+	dutyChan     chan Duty
+	deadlineChan chan Duty
+
+	// duties represents a set since duty deduplication is required.
+	duties map[Duty]bool
+}
+
+// NewDeadliner returns a new instance of Deadline.
+func NewDeadliner(ctx context.Context, eth2Cl slotTimeProvider) (*Deadline, error) {
+	genesis, err := eth2Cl.GenesisTime(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	duration, err := eth2Cl.SlotDuration(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Deadline{
+		genesisTime:  genesis,
+		slotDuration: duration,
+		dutyChan:     make(chan Duty),
+		deadlineChan: make(chan Duty),
+		duties:       make(map[Duty]bool),
+	}
+
+	go func() {
+		var (
+			deadlineTimer <-chan time.Time
+			minDuty       Duty
+		)
+
+		for {
+			select {
+			case <-ctx.Done():
+				// Killed by parent context
+				return
+			case duty := <-d.dutyChan:
+				if minDuty.Type == DutyUnknown {
```

Comment:
why do you assume that the first duty to get into `dutyChan` would be a `DutyUnknown`?

Reply:
here i am checking minDuty for its zero value to initialise minDuty with first duty obtained